src/lib.rs

```diff
@@ -884,6 +921,215 @@ pub unsafe trait FromBytes: FromZeroes {
+    ///
+    ///   Pragmatically, this means that if `*candidate` is guaranteed to
+    ///   contain an enum type at a particular offset, and the enum discriminant
+    ///   stored in `*candidate` corresponds to a valid variant of that enum
+    ///   type, then it is guaranteed that the appropriate bytes of `*candidate`
+    ///   are initialized as defined by that variant's bit validity (although
+    ///   note that the variant may contain another enum type, in which case the
+    ///   same rules apply depending on the state of its discriminant, and so on
+    ///   recursively).
+    ///
+    /// ## Postconditions
+    ///
+    /// Unsafe code may assume that, if `is_bit_valid(candidate)` returns true,
+    /// `*candidate` contains a valid `Self`.
+    ///
+    /// [valid]: https://doc.rust-lang.org/std/ptr/index.html#safety
+    #[doc(hidden)]
+    unsafe fn is_bit_valid(candidate: NonNull<Self>) -> bool;
+
+    /// Attempts to interpret a byte slice as a `Self`.
+    ///
+    /// `try_from_ref` validates that `bytes` contains a valid `Self` as defined
+    /// by [`is_bit_valid`]. If it does, then `bytes` is reinterpreted as a
+    /// `Self`.
+    ///
+    /// [`is_bit_valid`]: TryFromBytes::is_bit_valid
+    // TODO(#251): In a future in which we distinguish between `FromBytes` and
+    // `RefFromBytes`, this requires `where Self: RefFromBytes` to disallow
+    // interior mutability.
+    #[inline]
+    fn try_from_ref(bytes: &[u8]) -> Option<&Self> {
+        // PANICS: `bytes` is a `&[u8]`, and so the sum of its address and
+        // length cannot overflow `usize`.
+        let elems = Self::LAYOUT.validate_exact_cast(bytes.addr(), bytes.len())?;
+        let slf = Self::raw_from_ptr_len(NonNull::from(bytes).cast::<u8>(), elems);
+
+        // SAFETY:
+        // - `bytes` is a `&[u8]`, which guarantees that its length doesn't
+        //   overflow `isize` and that it comes from a single allocation
+        // - `validate_exact_cast` checked alignment
+        // - all bytes of `bytes` are initialized
+        if unsafe { !Self::is_bit_valid(slf) } {
+            return None;
+        }
+
+        // SAFETY:
+        // - `is_bit_valid` guaranteed that `*slf` contains a valid `Self`
+        // - Since `Self` is not allowed to contain any `UnsafeCell`s:
+        //   - The caller cannot use the `&Self` to perform interior mutation on
+        //     a byte range that `bytes` views as not containing `UnsafeCell`s
+        //   - The caller cannot use the `&Self` to write invalid values to
+        //     `bytes` (namely, uninitialized bytes, as `[u8]` has no other bit
+        //     validity constraints)
+        // - Since `[u8]` does not contain any `UnsafeCell`s, we are guaranteed
+        //   that, having verified that `slf` currently contains a valid `Self`,
+        //   code with access to `bytes` cannot cause it to no longer contain a
+        //   valid `Self` in the future
```

Comment:
This doesn't affect us right now: We only permit transmuting from `&[u8]`. There is no memory model under which it would be sound for us to allow `&[u8]` -> `&T` where `T` contains `UnsafeCell`s.* If we support `&T` -> `&U` at some point in the future, then this will become relevant.

\* In theory, this could be valid in a world where UnsafeCells only cause problems when _used_, but even then, it would only be valid for types which we define and so can guarantee never allow the interior mutability to be exercised.